### PR TITLE
Add new scheduling directive clone_in

### PR DESCRIPTION
**Pull request by psuriana**
Similar to `Func::in()`, but instead of creating an identity wrapper of the Func, it creates a clone of the Func. Consider the following code:
```
f(x) = x;
g(x) = f(x) + 2;
h(x) = f(x) * 2;
```
`f.clone_in(g)` will rewrite the pipeline into:
```
f(x) = x;
f_clone(x) = x;
g(x) = f_clone(x) + 2;
h(x) = f(x) * 2;
```

### CODE REVIEW

**Comment by abadams** [comment]
Addresses #2265 which has come up several times now.

**Comment by steven-johnson** [comment]
Given the similarity, it would probably be worth adding documentation giving usage examples of both in() and clone_in().

**Comment by psuriana** [comment]
@abadams PTAL

File: tutorial/lesson_20_cloning_funcs.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        //     f(x, y) = x + y
+        // for y:
+        //   for x:
+        //     g(x, y) = 2 * f(x, y) + 3
+        // for y:
+        //   for x:
+        //     h(x, y) = f(x, y) + g(x, y) + 10
+
+        // Using Func::clone, we can replace calls to 'f' inside 'g' with
+        // clone of 'f' using the schedule alone:
```
Comment (abadams) [inline tutorial/lesson_20_cloning_funcs.cpp]: "clone" -> "a clone"

File: tutorial/lesson_20_cloning_funcs.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        // for y:
+        //   for x:
+        //     out(x, y) = f(x, y) + g(x, y) + h(x, y)
+
+        out.realize(5, 5);
+    }
+
+    printf("Success!\n");
+
+    return 0;
```
Comment (abadams) [inline tutorial/lesson_20_cloning_funcs.cpp]: The thing that's missing here is an example of a realistic use-case. I guess one might be if two consumers of a Func need regions of it that are disjoint and far apart, then you'd be better of with two different copies of the producer.

File: tutorial/lesson_20_cloning_funcs.cpp [hunk 3]
```diff
@@ -137,6 +137,41 @@ int main(int argc, char **argv) {
...
+        // consume regions of the producer that are very disjoint. Consider
+        // the following case for example:
+        Func f("f"), g("g"), h("h");
+        f(x) = x;
+        g(x) = 2 * f(0);
+        h(x) = f(99) + 10;
+
+        // Let's schedule 'f' to be computed at root.
+        f.compute_root();
+        // Since both 'g' and 'h' consumes 'f', the region required of 'f'
```
Comment (abadams) [inline tutorial/lesson_20_cloning_funcs.cpp]: consumes -> consume

File: tutorial/lesson_20_cloning_funcs.cpp [hunk 4]
```diff
@@ -137,6 +137,41 @@ int main(int argc, char **argv) {
...
+        // in the x-dimension is [0, 99]. The equivalent loop nests are:
+        // for x = 0 to 99
+        //   f(x) = x
+        // for x:
+        //   g(x) = 2 * f(0)
+        // for x:
+        //   h(x) = f(99) + 10
+
+        // If 'f' had been very expensive to compute, we might be beter off
```
Comment (abadams) [inline tutorial/lesson_20_cloning_funcs.cpp]: had been -> is

Comment (abadams) [reply tutorial/lesson_20_cloning_funcs.cpp]: beter -> better

File: tutorial/lesson_20_cloning_funcs.cpp [hunk 5]
```diff
@@ -137,6 +137,41 @@ int main(int argc, char **argv) {
...
+        // with having two copies of 'f' for each consumer, 'g' and 'h', to
```
Comment (abadams) [inline tutorial/lesson_20_cloning_funcs.cpp]: two -> distinct

**Comment by psuriana** [comment]
@steven-johnson PTAL

File: tutorial/lesson_20_cloning_funcs.cpp [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+#include <stdio.h>
+
+using namespace Halide;
+
+int main(int argc, char **argv) {
+    // First we'll declare some Vars to use below.
+    Var x("x"), y("y"), xo("xo"), yo("yo"), xi("xi"), yi("yi");
+
+    // This lesson will be about cloning a Func using the Func::clone_in
+    // directive
```
Comment (steven-johnson) [inline tutorial/lesson_20_cloning_funcs.cpp]: Add a . to the end of the sentence

File: tutorial/lesson_20_cloning_funcs.cpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        //   for x:
+        //     f(x, y) = x + y
+        // for y:
+        //   for x:
+        //     g(x, y) = 2 * f(x, y) + 3
+        // for y:
+        //   for x:
+        //     h(x, y) = f(x, y) + g(x, y) + 10
+
+        // Using Func::clone, we can replace calls to 'f' inside 'g' with
```
Comment (steven-johnson) [inline tutorial/lesson_20_cloning_funcs.cpp]: Func::clone_in

**Pull request merged**